updater/aws_test.go

```diff
@@ -15,6 +15,244 @@ import (
+		StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateAvailable, "v1.0.5")),
+	}
+
+	for i := 0; i < 100; i++ {
+		containerID := "cont-inst-br" + strconv.Itoa(i)
+		ec2ID := "ec2-id-br" + strconv.Itoa(i)
+		instances = append(instances, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+		})
+		expected = append(expected, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+			bottlerocketVersion: "v1.0.5",
+		})
+	}
+
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return getOut, nil
+		},
+		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
+			return &ssm.SendCommandOutput{
+				Command: &ssm.Command{
+					CommandId:    aws.String("command-id"),
+					DocumentName: aws.String("check-document"),
+				},
+			}, nil
+		},
+		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
+			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
+			return nil
+		},
+	}
+	u := updater{ssm: mockSSM}
+	actual, err := u.filterAvailableUpdates(instances)
+	require.NoError(t, err)
+	assert.EqualValues(t, expected, actual, "should contain all instances")
+}
+
+func TestPaginatedFilterAvailableUpdatesAllFail(t *testing.T) {
+	instances := make([]instance, 0)
+
+	for i := 0; i < 100; i++ {
+		containerID := "cont-inst-br" + strconv.Itoa(i)
+		ec2ID := "ec2-id-br" + strconv.Itoa(i)
+		instances = append(instances, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+		})
+	}
+
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return nil, nil
+		},
+		SendCommandFn: func(input *ssm.SendCommandInput) (*ssm.SendCommandOutput, error) {
+			return nil, errors.New("Failed to send document")
+		},
+		WaitUntilCommandExecutedWithContextFn: func(ctx aws.Context, input *ssm.GetCommandInvocationInput, opts ...request.WaiterOption) error {
+			assert.Equal(t, "command-id", aws.StringValue(input.CommandId))
+			return nil
+		},
+	}
+	u := updater{ssm: mockSSM}
+	actual, err := u.filterAvailableUpdates(instances)
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "Failed to send document")
+	assert.Empty(t, actual)
+}
+
+func TestPaginatedFilterAvailableUpdatesWithinnPageAllFail(t *testing.T) {
+	instances := make([]instance, 0)
+
+	for i := 0; i < 100; i++ {
+		containerID := "cont-inst-br" + strconv.Itoa(i)
+		ec2ID := "ec2-id-br" + strconv.Itoa(i)
+		instances = append(instances, instance{
+			instanceID:          ec2ID,
+			containerInstanceID: containerID,
+		})
+	}
+
+	mockSSM := MockSSM{
+		GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
+			return nil, errors.New("Failed to get command output")
```

Comment:
Instead of failing for all the instances in a page can we do mix of failures so that all the lines [here](https://github.com/bottlerocket-os/bottlerocket-ecs-updater/blob/develop/updater/aws.go#L186-L199) are tested. We can do something like:
```
count = 0
GetCommandInvocationFn: func(input *ssm.GetCommandInvocationInput) (*ssm.GetCommandInvocationOutput, error) {
        count++	
        swtich count%3 {
	case 0: 
               return nil, errors.New("Failed to get command output") // validates getCommandResult failure
	case 1: 
               return "{}", nil // validates parseCommandOutput failure
	case 2: 
               return &ssm.GetCommandInvocationOutput{
		         Status:                aws.String("Success"),
		         StandardOutputContent: aws.String(fmt.Sprintf(checkPattern, updateStateAvailable, "v1.0.5")),
	        }, nil // validates success case
	}
}
```

Reply:
I had considered doing something like that, but since both `parseCommandOutput` and `GetCommandInvocation` were `continue` on fail it seemed redundant to test both since the behavior would be the same.

Reply:
I'll try that out. I had considered doing something along these lines initially but thought it might be a bit redundant since the behavior for `parseCommandOutput` and `getCommandInvocation`'s error handling are the same. But having a success case in there with the failures is a good call out. Thanks!

Reply:
It can be redundant behavior, but it is important that all the lines and conditions are touched by the tests to make sure any updates to the method is validated by the tests.
For assertion you can just assert instance count instead of matching every 3rd container instance id.